File: LeetCode/Depth-First Search/Easy/733. Flood Fill/main.py
"""
In this flood filled algo we color the cells adjacent to our cell . 
We can either keep a visited set , of tupples(i,j) for returning if we goto same block which is frequently used in most bfs algo 
but here since we have constrain that color==new_color(avoid going to predifend cells of new color in the question iteself) 
to avoid iterting through already colored elements  

"""


class Solution:
    def floodFill(self, image, sr, sc, newColor):

        h, w = len(image), len(image[0])
        # visited = set()       # not using this here

        def dfs(image, r, c, color, new_color):
            # makign some constraint to avoid it going overflow or iterating already iterated through or block we dont wann iterate
            if(r < 0 or c < 0 or r >= h or c >= w or color == new_color or image[r][c] != color):
                return

            # filling the color in the matrix
            image[r][c] = new_color

            # flooding into adjacent cells
            dfs(image, r+1, c, color, new_color)
            dfs(image, r-1, c, color, new_color)
            dfs(image, r, c+1, color, new_color)
            dfs(image, r, c-1, color, new_color)

        dfs(image, sr, sc, image[sr][sc], newColor)
        return image


# testing
s = Solution()
print(s.floodFill([[1, 1, 1], [1, 1, 0], [1, 0, 1]], 1, 1, 2))

File: LeetCode/Depth-First Search/Easy/733. Flood Fill/main.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc, int newColor)
    {

        dfs(image, sr, sc, image[sr][sc], newColor);
        return image;
    }

    void dfs(vector<vector<int>> &image, int r, int c, int color, int newColor)
    {

        if (r < 0 || c < 0 || r >= image.size() || c >= image[0].size() || image[r][c] == newColor || image[r][c] != color)
            return;

        image[r][c] = newColor;

        dfs(image, r + 1, c, color, newColor);
        dfs(image, r - 1, c, color, newColor);
        dfs(image, r, c + 1, color, newColor);
        dfs(image, r, c - 1, color, newColor);
    }
};

// testing
int main()
{
    Solution obj;
    vector<vector<int>>
        input{{1, 1, 1}, {1, 1, 0}, {1, 0, 1}};
    vector<vector<int>> ans;
    ans = obj.floodFill(input, 1, 1, 2);
    for (int i = 0; i < ans.size(); i++)
    {
        for (int j = 0; j < ans[i].size(); j++)
        {
            cout << ans[i][j] << " ";
        }
        cout << endl;
    }
}